src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedStream.cs

```diff
@@ -0,0 +1,1098 @@
+        {
+            try
+            {
+                if (disposing && _stream != null)
+                {
+                    try
+                    {
+                        Flush();
+                    }
+                    finally
+                    {
+                        _stream.Dispose();
+                    }
+                }
+            }
+            finally
+            {
+                _stream = null;
+                _buffer = null;
+                _lastSyncCompletedReadTask = null;
+
+                // Call base.Dispose(bool) to cleanup async IO resources
+                base.Dispose(disposing);
+            }
+        }
+
+
+        public override void Flush()
+        {
+            EnsureNotClosed();
+
+            // Has WRITE data in the buffer:
+            if (_writePos > 0)
+            {
+                FlushWrite();
+                Contract.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
+                return;
+            }
+
+            // Has READ data in the buffer:
+            if (_readPos < _readLen)
+            {
+                // If the underlying stream is not seekable AND we have something in the read buffer, then FlushRead would throw.
+                // We can either throw away the buffer resulting in data loss (!) or ignore the Flush.
+                // (We cannot throw becasue it would be a breaking change.) We opt into ignoring the Flush in that situation.
+                if (!_stream.CanSeek)
+                    return;
+
+                FlushRead();
+
+                // User streams may have opted to throw from Flush if CanWrite is false (although the abstract Stream does not do so).
+                // However, if we do not forward the Flush to the underlying stream, we may have problems when chaining several streams.
+                // Let us make a best effort attempt:
+                if (_stream.CanWrite || _stream is BufferedStream)
+                    _stream.Flush();
+
+                Contract.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
+                return;
+            }
+
+            // We had no data in the buffer, but we still need to tell the underlying stream to flush.
+            if (_stream.CanWrite || _stream is BufferedStream)
+                _stream.Flush();
+
+            _writePos = _readPos = _readLen = 0;
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled<int>(cancellationToken);
+
+            EnsureNotClosed();
+
+            return FlushAsyncInternal(cancellationToken, this, _stream, _writePos, _readPos, _readLen);
+        }
+
+
+        private static async Task FlushAsyncInternal(CancellationToken cancellationToken,
+                                                     BufferedStream _this, Stream stream, int writePos, int readPos,
```

Comment:
NIT: why `_this` as a parameter? Is it just intended as a hint for the developer?